--- system-macos26.sb	2026-09-26 00:55:08
+++ system-macos27.sb	2026-09-26 00:55:08
@@ -59,6 +59,7 @@
   (allow file-read* file-test-existence file-map-executable
          (subpath "/AppleInternal/Developer/Library/Frameworks")
          (subpath "/AppleInternal/Developer/usr/lib")
+         (subpath "/AppleInternal/Library/Extensions")
          (subpath "/AppleInternal/Library/Frameworks")
          (subpath "/AppleInternal/Library/PrivateFrameworks")
          (subpath "/AppleInternal/usr/lib")
@@ -163,7 +164,6 @@
        (global-name "com.apple.cfprefsd.daemon")
        (global-name "com.apple.diagnosticd")
        (global-name "com.apple.dt.automationmode.reader")
-       (global-name "com.apple.espd")
        (global-name "com.apple.logd")
        (global-name "com.apple.logd.events")
        (global-name "com.apple.runningboard")
